feat(dragdrop): Add support of Drag and Drop (#4270)
unoplatform/uno: doc/articles/features/pointers-keyboard-and-other-user-inputs.md

> # Uno support for user inputs

Reviewer: Great doc! Should be added to TOC https://github.com/unoplatform/uno/blob/master/doc/articles/toc.yml

## Supported user inputs

User inputs are usually propagated to the using `RoutedEvent`, see. Uno's [routed events documentation](routed-events.md) to better understand their implementation on Uno.

> | Routed Event                  | Android | iOS     | Wasm    | MacOS | Skia WPF | Skia GtK | Tizen |    |

Reviewer 2: I think there is a mismatch between the table header and column count. The table doesn't render correctly to HTML.

| _drag and drop_
| `DragStarting`                | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.dragstarting) |
| `DragEnter`                   | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.dragenter) |
| `DragOver`                    | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.dragover) |
| `DragLeave`                   | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.dragleave) |
| `Drop`                        | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.drop) |
| `DropCompleted`               | Yes     | Yes     | Yes     | Yes   | Yes      | Yes      | Yes   | [Documentation](https://docs.microsoft.com/uwp/api/windows.ui.xaml.uielement.dropcompleted) |

Notes:

> 1. **Focus** events:

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

>    * **iOS**: The concept of _focus_ is emulated because not supported by the platform, so this event is

Reviewer: ```suggestion
   * **iOS**: The concept of _focus_ is emulated because it's not supported by the platform, so this event is
```

>      always bubbling in managed code.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

>    * **Wasm**: Current implementation is not totally reliable and doesn't support _lost focus_ most of the time.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> 2. **Keyboard** events:

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

>    * **Android**: `KeyDown` and `KeyUp` events are **generated only from hardware keyboards** (Except for the _Editor Action_ on _soft keyboards_,

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

     those are translated as `KeyUp` with `KeyCode.Enter`). Some soft keyboard **MAY** generate those events, but your code shouldn't rely
     on that. This is a limitation [in the Android platform](https://developer.android.com/training/keyboard-input/commands) (see note on this link content).
     > Because of those limitations, _Key Events_ are not being implemented as _routed events_ on Android, so `AddHandler` & `RemoveHandler`

>      > won't work for keyboard events. **They won't bubble in managed code**.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

   * **iOS**: `KeyDown` & `KeyUp` routed events are generated from only a `TextBox`. Only character-related keyboard events are generated.
     They are implemented as _Routed Events_ and they are **always bubbling in managed code**.

## Pointer Events

> Those events are the base for all others pointing device related events (i.e. Manipulation, Gesture and drag and dop events).

Reviewer: ```suggestion
These events are the base for all other pointing device related events (i.e. Manipulation, Gesture and drag and dop events).
```

They are directly linked to the native events of each platform:

> * `Touches[Began|Moved|Ended|Cancelled]` on iOS

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

* `dispatchTouchEvent` and `dispatchGenericMotionEvent` on Android
* `pointer[enter|leave|down|up|move|cancel]` on WebAssembly

> On skia however, they are fully managed events.

Reviewer: ```suggestion
On Skia however, they are fully managed events.
```

### Pointers events and the ScrollViewer

Like on WinUI as soon as the system detects that the user wants to scroll, a control gets a `PointerCancelled` and that control won't receive

> any other pointer event until the user release the pointer. That behavior can be prevented by setting the `ManipulationMode` 

Reviewer: ```suggestion
any other pointer event until the user releases the pointer. That behavior can be prevented by setting the `ManipulationMode` 
```

to something else than `System` on a control nested in the `ScrollViewer`. (cf. [Manipulation events](#Manipulation_Events))

Be aware that on iOS this will set `DelaysContentTouches` to `false` so it means that it will slightly reduce the performance
of the scrolling (cf. [documentation](https://developer.apple.com/documentation/uikit/uiscrollview/1619398-delayscontenttouches)).

### Known limitations for pointer events

As those events are tightly coupled to the native events, Uno has to make some compromises:
* On iOS, when tapping with a mouse or a pen on Android, or in few other specific cases (like `PointerCaptureLost`), 
  multiple managed events are raised from a single native event. These have multiple effects:
	* On UWP if you have a control A and a nested control B, you will get:

> 		```

Reviewer 3: Codacy found an issue: [[list-item-content-indent] Don’t use mixed indentation for children, add 1 space](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

		B.PointerEnter
		A.PointerEnter
		B.PointerPressed
		A.PointerPressed
		```
	  but with UNO you will get:
		```
		B.PointerEnter
		B.PointerPressed
		A.PointerEnter
		A.PointerPressed

> 		```

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

	* If you handle the `PointerEnter` on **B**, the parent control **A** won't get the `PointerEnter` (as expected) nor the  `PointerPressed`.
* On Android with a mouse or a pen, the `PointerEnter` and `PointerExit` are going to be raised without taking clipping in consideration.
  This means that you will get the enter earlier and the exit later than on other platform.
* On Android if you have an element with a `RenderTransform` which overlaps one of its sibling element, 
  the element at the top will get the pointer events.

> * On WASM, iOS and Android, the `RoutedPointerEventArgs.FrameId` will be reset to 0 after 49 days of running time of the application.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * On WASM, iOS and Android, the `RoutedPointerEventArgs.FrameId` will be reset to 0 after 49 days of running time of the application.

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * Unlike on UWP, controls that are under a `Popup` won't receive the unhandled pointer events.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * On non skia-based platforms, unlike UWP, it's impossible to receive a `PointerReleased` without getting a `PointerPressed` before.

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * On non skia-based platforms, unlike UWP, it's impossible to receive a `PointerReleased` without getting a `PointerPressed` before.

Reviewer: ```suggestion
* On non Skia-based platforms, unlike UWP, it's impossible to receive a `PointerReleased` without getting a `PointerPressed` before.
```

  (For instance if a child control handled the pressed event but not the released event).

>   > On WASM as `TextElement` are `UIElement`, it means that unlike UWP `TextBlock` won't raise the 

Reviewer: ```suggestion
  > On WASM as `TextElement` inherits from `UIElement`, it means that unlike UWP `TextBlock` won't raise the 
```

>   > a `PointerReleased` when clicking on an `Hyperlink`.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

>   > a `PointerReleased` when clicking on an `Hyperlink`.

Reviewer: ```suggestion
  > `PointerReleased` event when clicking on a `Hyperlink`.
```

> * Unlike UWP, on the `Hyperlink` the `Click` will be raised before the `PointerReleased`.

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * Unlike UWP, on the `Hyperlink` the `Click` will be raised before the `PointerReleased`.

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

* The property `PointerPointProperties.PointerUpdateKind` is not set on Android 5.x and lower (API level < 23)
* On Firefox, pressed pointers are reported as fingers. This means you will receive events with `PointerDeviceType == Pen` only for hovering 
  (i.e. `Pointer<Enter|Move|Exit>` - note that, as of 2019-11-28, once pressed `PointerMove` will be flagged as "touch") 

>   and you won't  be able to track the barrel button nor the eraser. (cf. https://bugzilla.mozilla.org/show_bug.cgi?id=1449660)

Reviewer 3: Codacy found an issue: [[no-literal-urls] Don’t use literal URLs without angle brackets](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

>   and you won't  be able to track the barrel button nor the eraser. (cf. https://bugzilla.mozilla.org/show_bug.cgi?id=1449660)

Reviewer 3: Codacy found an issue: [[list-item-spacing] Missing new line after list item](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

> * On WASM, if you touch the screen with the pen **then** you press the barrel button (still while touching the screen), the pointer events will

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

  have the `IsRightButtonPressed` set (in addition of the `IsBarrelButtonPressed`). On WinUI and Android you get this flag only if the barrel 
  button was pressed at the moment where you touched the screen, otherwise you will have the `IsLeftButtonPressed` and the `IsBarrelButtonPressed`.

> * For pen and fingers, the `Holding` event is not raised after a given delay like on WinUI, 

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

  but instead we rely on the fact that we usually get a lot of moves for those kind of pointers, 
  so we raise the event only when we get a move that exceed defined thresholds for holding.
* On WASM, Shapes must have a non null Fill to receive pointer events (setting the Stroke is not sufficient).

> * On WASM if the user scrolls in diagonal (e.g. with a Touchpad), but you mark as `Handled` pointer events only for vertical scrolling,

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)

  then the events for the horizontal scroll component won't bubble through the parents.

### Pointer capture

The capture of pointer is handled in managed code only. On WebAssembly Uno however still requests the browser to capture the pointer,
but Uno does not rely on native `[got|lost]pointercapture` events.

### iPadOS mouse support

To differentiate between mouse and touch device type for pointer events, include the following in your app's `Info.plist`:

```
<key>UIApplicationSupportsIndirectInputEvents</key>
<true/>
```

Without this key the current version of iPadOS reports mouse interaction as normal touch.

## Manipulation Events

They are generated from the PointerXXX events (using the `Windows.UI.Input.GestureRecognizer`) and are bubbling in managed only.

> Currently there is no intertia support, so the `IsInertial` will always be `false` and the `UIElement.ManipulationInertiaStarting` event 

Reviewer: ```suggestion
Currently there is no inertia support, so the `IsInertial` will always be `false` and the `UIElement.ManipulationInertiaStarting` event 
```

> will never be fired. The `Velocities` properties of event args are not implemented neither.

Reviewer: ```suggestion
will never be fired. The `Velocities` properties of event args are not implemented either.
```

## Gesture Events

They are generated from the PointerXXX events (using the `Windows.UI.Input.GestureRecognizer`) and are bubbling in managed only.

Note that `Tapped` and `DoubleTapped` are not linked in any way to a native equivalent, but are fully interpreted in managed code.

In order to match the WinUI behavior, on WASM the default "Context menu" of the browser is disabled (except for the `TextBox`), 
no matter if you use / handle the `RightTapped` event or not.
Be aware that on some browser (Firefox), user can still request to get the "Context menu" on right click.

## Drag and drop

Those events are also 100% managed events, built from the PointerXXX events (using the `Windows.UI.Input.GestureRecognizer`)

### Intra-app drag and drop support

A _drag and drop_ operation can be used to move content within an app, but it can also be used to **copy** / **move** / **link** between apps.

> While inter-app _drag and drop_ is supported on all platforms without limiations, intra-app _drag and drop_ requires platfrom specific support.

Reviewer: ```suggestion
While inter-app _drag and drop_ is supported on all platforms without limitations, intra-app _drag and drop_ requires platform specific support.
```

I corrected the typos but... is it true? You write "supported on all platforms without limitations" but there's a lot of 'No' below :p

Author: Yeah it's true: the "inter-app" does not suffer real limitation, just a minor bug which won't have any functional impact.

Reviewer: Ok then are you mixing up 'inter-' and 'intra-'?

Author: Whoups rigth! ... Fixed

| -------- | ------------------------------------------------ | ------------------------                        |
| Android  | No _in progress_                                 | No _in progress_                                |
| iOS      | No                                               | No                                              |
| Wasm     | No                                               | No                                              |
| MacOS    | No _in progress_                                 | No _in progress_                                |
| Skia WPF | Yes (Text, Link (1), Image (2), File, Html, Rtf) | Yes (Text, WebLink (1), Image, File, Html, Rtf) |
| Skia GTK | No                                               | No                                              |

1. There is not standard type for **WebLink** (nor **ApplicationLink**) on this platform. 
   They are copied to the external app as raw **Text**, and converted back as **WebLink** from raw text from the external app 
   when [`Uri.IsWellFormedUriString(text, UriKind.Absolute)](https://docs.microsoft.com/en-us/dotnet/api/system.uri.iswellformeduristring) returns true.

> 2. The image content seems to not being readable by common apps, only another Uno app is able to read it properly.

Reviewer 3: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)
